enhancement(http provider): Add automatic bearer token acquisition in http-client [version 2]

KowalczykBartek opened the pull request:
PR to solve a https://github.com/vectordotdev/vector/issues/20635 feature request.
This change brings new section in configuration for HttpSinkConfig :
```rust
pub struct HttpClientAuthorizationConfig {
    auth: HttpClientAuthorizationStrategy,
    tls: Option<TlsConfig>,
}
```
this allows to explicit define a oauth2 server which will be used to acquire bearer token (also, this request allows to bring mtls via TlsConfig),
```yaml
sources:
  my_source_id:
    scrape_interval_secs: 1
    type: http_client
    endpoint: http://127.0.0.1:9898/logs

sinks:
  my_sink_id:
    http_client_authorization_strategy:
      auth:
        strategy: "o_auth2"
        client_id: "your.clientid"
        client_secret: "verysecretstring"
        token_endpoint: "https://your.provider.com/oauth/token"
    type: http
    encoding:
      codec: raw_message
    inputs:
      - my_source_id
    uri: http://127.0.0.1:9091/output
```
Some basic error handling is also on place, so invalid credentials will not crash a vector instance :) you can expect an event and error message with body from oauth server (unfortunately I over engineered this error handling a little bit, I have limited rust knowledge and was unable to fully understand all patterns - so, suggestions, how to improve this would be appreciated :) )
```
2024-10-22T21:51:18.256079Z  WARN sink{component_kind="sink" component_id=my_sink_id component_type=http}:request{request_id=1}:http: vector::internal_events::http_client: HTTP Auth extension error. error=Server error from authentication server: {"error":"invalid_client","error_description":"Bad credentials"} error_type="request_failed" stage="processing" internal_log_rate_limit=true
```

So, as discussed in https://github.com/vectordotdev/vector/pull/21023#discussion_r1792213783 - @pront I would be grateful for review :)

KowalczykBartek commented on src/http.rs:
```diff
@@ -119,13 +300,32 @@ where
+            //should request for token influence upstream service latency ?            
+            if let Some(auth_extension) = auth_extension {
+                let auth_span = tracing::info_span!("auth_extension");
+                let res = auth_extension.modify_request(&mut request)
```
I had no idea how to handle this Box<dyn Error> to some vector friendly error better :/

KowalczykBartek replied:
I was able to improve it a little bit with snafu 
```rust
auth_extension.modify_request(&mut request)
    .instrument(auth_span.clone().or_current())
    .await
    .inspect_err(|error| {
        // Emit the error into the internal events system.
        emit!(http_client::GotAuthExtensionError{error});
    })
    .context(ExtensionAuthenticationSnafu)?;
```

KowalczykBartek commented on src/http.rs:
```diff
@@ -72,31 +75,208 @@ impl HttpError {
 pub type HttpClientFuture = <HttpClient as Service<http::Request<Body>>>::Future;
 type HttpProxyConnector = ProxyConnector<HttpsConnector<HttpConnector>>;
 
+#[async_trait]
```
I wanted to do it more in hyper's Service manner but I failed :D

KowalczykBartek replied:
fixed.

pront commented on src/http.rs:
```diff
@@ -56,6 +56,10 @@ pub enum HttpError {
     #[snafu(display("Failed to build HTTP request: {}", source))]
     BuildRequest { source: http::Error },
+    #[snafu(display("Failed to acquire authentication resource."))]
+    ExtensionAuthentication {
```
Nit: rename to `AuthenticationExtension`.

KowalczykBartek replied:
fixed

pront commented on src/http.rs:
```diff
@@ -262,6 +499,65 @@ impl<B> fmt::Debug for HttpClient<B> {
+pub struct HttpClientAuthorizationConfig {
+    /// Define how to authorize against an upstream.
+    #[configurable]
+    auth: HttpClientAuthorizationStrategy,
```
Can we flatten this to avoid repetition in configs, from: 
```
    http_client_authorization_strategy:
      auth:
        strategy: "o_auth2"
        client_id: "your.clientid"
        client_secret: "verysecretstring"
        token_endpoint: "https://your.provider.com/oauth/token"
```
to
```
    http_client_authorization_strategy:
      strategy: "o_auth2"
      client_id: "your.clientid"
      client_secret: "verysecretstring"
      token_endpoint: "https://your.provider.com/oauth/token"
```

I think this makes sense from a UX perspective cc @vectordotdev/ux-team

KowalczykBartek replied:
Honestly I was unable to figure out this - @pront  can you pls point me some example with flattened enum ?

pront replied:
I took another look at this:
* We can replace `pub http_client_authorization_strategy: Option<HttpClientAuthorizationConfig>` with `pub authorization_config: Option<AuthorizationConfig>`
* Rename `auth` to `strategy`

KowalczykBartek replied:
then we end with double strategy.strategy, is that ok ?
```yaml
sinks:
  my_sink_id:
    authorization_config:
      strategy:
        strategy: "o_auth2"
        client_id: "some.client.id"
        client_secret: "ab!@#"
        # token_endpoint: "https://profile.api.stage.context.cloud.sap/uaa/oauth/token"
        grace_period: 780
        token_endpoint: "https://127.0.0.1:8091/oauth/token"
```
maybe I will change this tag 
```yaml
/// Configuration of the authentication strategy for HTTP requests.
///
/// HTTP authentication should be used with HTTPS only, as the authentication credentials are passed as an
/// HTTP header without any additional encryption beyond what is provided by the transport itself.
#[configurable_component]
#[derive(Clone, Debug)]
#[serde(deny_unknown_fields, rename_all = "snake_case", tag = "strategy")]
#[configurable(metadata(docs::enum_tag_description = "The authentication strategy to use."))]
```
to 'protocol' ?

KowalczykBartek replied:
then config will look like this 
```yaml
sinks:
  my_sink_id:
    authorization_config:
      strategy:
        strategy: "o_auth2"
        client_id: "some.client.id"
        client_secret: "ab!@#"
        grace_period: 780
        token_endpoint: "https://127.0.0.1:8091/oauth/token"
```
is this double strategy ok ? maybe I will change also tag from 
```yaml
/// Configuration of the authentication strategy for HTTP requests.
///
/// HTTP authentication should be used with HTTPS only, as the authentication credentials are passed as an
/// HTTP header without any additional encryption beyond what is provided by the transport itself.
#[configurable_component]
#[derive(Clone, Debug)]
#[serde(deny_unknown_fields, rename_all = "snake_case", tag = "strategy")]
#[configurable(metadata(docs::enum_tag_description = "The authentication strategy to use."))]
```
to 'protocol'

KowalczykBartek replied:
together with fix for mtls extension I pushed renamed struct 
```
#[configurable(derived)]
pub authorization_config: Option<AuthorizationConfig>,
```

pront commented on src/internal_events/http_client.rs:
```diff
@@ -95,6 +95,24 @@ impl<'a> InternalEvent for GotHttpWarning<'a> {
+
+impl<'a> InternalEvent for GotAuthExtensionError<'a>  {
+    fn emit(self) {
+        warn!(
```
This should be `error!`.

KowalczykBartek replied:
fixed.

pront commented on src/internal_events/http_client.rs:
```diff
@@ -95,6 +95,24 @@ impl<'a> InternalEvent for GotHttpWarning<'a> {
+            stage = error_stage::PROCESSING,
+            internal_log_rate_limit = true,
+        );
+        counter!("http_auth_ext_client_errors_total", "error_kind" => self.error.to_string()).increment(1);
```
Here we can do the following:
```suggestion
                counter!(
            "component_errors_total",
            "error_type" => error_type::CONFIGURATION_FAILED,
            "stage" => error_stage::SENDING,
        )
        .increment(1);
```

KowalczykBartek replied:
applied.

pront commented on src/http.rs:
```diff
@@ -119,13 +301,28 @@ where
+                    .await
+                    .inspect_err(|error| {
+                        // Emit the error into the internal events system.
+                        emit!(http_client::GotAuthExtensionError{error});
```
Nit: rename to `AuthExtensionError`

KowalczykBartek replied:
fixed

pront reviewed (commented):
Hi @KowalczykBartek thank you for this contribution! 

I did a quick scan of the changes and left a few comments.

KowalczykBartek commented on .gitignore:
```diff
@@ -10,6 +10,7 @@ checkpoints/*
 miniodat
 bench_output.txt
 sample.log
+vector.yaml
```
fixed.

KowalczykBartek commented:
I introduced grace_period property so, it can be adjusted - before it was hard coded for 1min. I also played around this feature more and found bug around grace calculation (overflow), so also fixed this.
I also verified following config works (so use case from original issue, oauth2 + mtls)
```yaml
sinks:
  my_sink_id:
    http_client_authorization_strategy:
      auth:
        strategy: "o_auth2"
        client_id: "clientid"
        client_secret: "secret"
        grace_period: 300
        token_endpoint: "https://127.0.0.1:8091/uaa/oauth/token"
      tls:
        crt_path: /config/cert.pem
        key_file: /config/server.key
```

pront commented on src/sinks/http/config.rs:
```diff
@@ -90,6 +91,9 @@ pub struct HttpSinkConfig {
     pub tls: Option<TlsConfig>,
 
+    #[configurable(derived)]
+    pub http_client_authorization_strategy: Option<HttpClientAuthorizationConfig>,
```
```suggestion
    pub authorization_config: Option<AuthorizationConfig>,
```

pront commented on src/http.rs:
```diff
@@ -262,6 +510,79 @@ impl<B> fmt::Debug for HttpClient<B> {
+#[derive(Clone, Debug)]
+#[serde(deny_unknown_fields, rename_all = "snake_case", tag = "strategy")]
+#[configurable(metadata(docs::enum_tag_description = "The authentication strategy to use."))]
+pub enum HttpClientAuthorizationStrategy {
```
```suggestion
pub enum AuthorizationStrategy {
```

pront reviewed (commented):
This is starting to look great. Left a few nits, these should improve config readability.

KowalczykBartek commented:
after testing on real environments I found another issue, oauth2 + mtls should be strict to the efc https://datatracker.ietf.org/doc/html/rfc8705, so, speaking in cURL
this is oauth2
```bash
curl --cert cert.pem --key key.pem --request POST \
  --url https://something/oauth/token \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data 'grant_type=client_credentials' \
  --data 'client_id=abcdabcdabcd'
``` 
this is mtls extension
```bash
curl --request POST \
  --url https://something/oauth/token \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data client_secret=abcdabcdabcd \
  --data grant_type=client_credentials \
  --data client_id=abcdabcdabcd
```
I also updated a comment around HttpClientAuthorizationStrategy - I hope this is correctly used rustdocs

pront reviewed (commented):
👋 Left some comments. Most of them are nits but we should get rid of the unwraps. 

Also, we lack tests here. We could mock the endpoint but we can also test some internal bits separately. Please see existing tests for inspiration. 

Thank you again for this contribution!

KowalczykBartek commented on src/http.rs:
```diff
@@ -72,31 +77,224 @@ impl HttpError {
+        //but later we operate on milliseconds. 
+        let now = SystemTime::now();
+        let since_the_epoch = now.duration_since(UNIX_EPOCH).unwrap();
+        let token_will_expire_after_ms = since_the_epoch.as_millis() + (token_is_valid_for_ms as u128);
```
fixed.

KowalczykBartek commented on src/http.rs:
```diff
@@ -72,31 +77,224 @@ impl HttpError {
+    B::Data: Send,
+    B::Error: Into<crate::Error> + Send,
+{
+    async fn modify_request(&self, req: &mut Request<B>) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
```
works out of the box 
```rust
async fn get_token(&self) -> Result<String, vector_lib::Error> 
```
fixed.

KowalczykBartek commented on src/http.rs:
````diff
@@ -262,6 +517,105 @@ impl<B> fmt::Debug for HttpClient<B> {
+    /// 
+    /// # Example
+    /// 
+    /// ```
````
fixed.

KowalczykBartek commented on src/http.rs:
````diff
@@ -262,6 +517,105 @@ impl<B> fmt::Debug for HttpClient<B> {
+    /// 
+    /// # Example
+    /// 
+    /// ```
````
fixed

KowalczykBartek commented on src/http.rs:
```diff
@@ -119,13 +318,28 @@ where
 
         default_request_headers(&mut request, &self.user_agent);
-        self.maybe_add_proxy_headers(&mut request);
+        self.maybe_add_proxy_headers(&mut request); 
```
fixed

KowalczykBartek commented on src/http.rs:
```diff
@@ -72,31 +77,224 @@ impl HttpError {
+    async fn request_token(&self) -> Result<ExpirableToken, Box<dyn std::error::Error + Send + Sync>> {
+        let mut request_body = format!("grant_type=client_credentials&client_id={}", self.client_id);
+        
+        //in case of oauth2 with mTLS (https://datatracker.ietf.org/doc/html/rfc8705) we only pass client_id,
```
fixed

KowalczykBartek commented on src/http.rs:
```diff
@@ -72,31 +77,224 @@ impl HttpError {
+        //we are multiplying by 1000 because expires_in field is in seconds, grace_period also, 
+        //but later we operate on milliseconds. 
+        let now = SystemTime::now();
+        let since_the_epoch = now.duration_since(UNIX_EPOCH).unwrap();
```
I handled time like this
```rust
let since_the_epoch = now.duration_since(UNIX_EPOCH).expect("time went backwards");
```
I saw this in vector, in this part 
```rust
let now = SystemTime::now()
    .duration_since(SystemTime::UNIX_EPOCH)
    .expect("Time can't drift behind the epoch!")
    .as_millis();
```

KowalczykBartek commented:
> Also, we lack tests here. We could mock the endpoint but we can also test some internal bits separately. Please see existing tests for inspiration.

just FYI (because I see some jobs running :D ) - i will provide tests, I just need few days more :)

pront commented:
> > Also, we lack tests here. We could mock the endpoint but we can also test some internal bits separately. Please see existing tests for inspiration.
> 
> just FYI (because I see some jobs running :D ) - i will provide tests, I just need few days more :)

Thank you @KowalczykBartek, much appreciated. Also, I wanted to share a few notes to help with the CI checks:
* we can add an `http` scope [here](https://github.com/vectordotdev/vector/blob/master/.github/workflows/semantic.yml#L31)
* also see our [changelog guide](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md)
* `cargo install -f --path vdev #cd to the repo root` 
* Run `cargo vdev fmt`
* Run `cargo vdev check rust`

KowalczykBartek commented on src/http.rs:
```diff
@@ -72,31 +77,224 @@ impl HttpError {
+        //but later we operate on milliseconds. 
+        let now = SystemTime::now();
+        let since_the_epoch = now.duration_since(UNIX_EPOCH).unwrap();
+        let token_will_expire_after_ms = since_the_epoch.as_millis() + (token_is_valid_for_ms as u128);
```
fixed.

KowalczykBartek replied:
fixed.

KowalczykBartek commented:
@pront missing bit is:
```
cargo vdev check rust --clippy
```
this command gives me some incorrect solution for error it shows,
```
error: you seem to be trying to use `&Box<T>`. Consider using just `&T`
   --> src/internal_events/http_client.rs:100:16
    |
100 |     pub error: &'a Box<dyn std::error::Error + Send + Sync>,
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&'a (dyn std::error::Error + Send + Sync)`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrowed_box
```
[,let's try](help: try: `&'a (dyn std::error::Error + Send + Sync)`) - ok lets try
```rust
#[derive(Debug)]
pub struct AuthExtensionError<'a> {
    pub error: &'a (dyn std::error::Error + Send + Sync),
}
```
doesn't work
```rust
error[E0277]: the size for values of type `dyn snafu::Error + std::marker::Send + std::marker::Sync` cannot be known at compilation time
   --> src/http.rs:416:65
    |
416 |                         emit!(http_client::AuthExtensionError { error });
    |                                                                 ^^^^^ doesn't have a size known at compile-time
    |
```
honestly I don't have idea how to solve this

pront commented:
> @pront missing bit is:
> 
> ```
> cargo vdev check rust --clippy
> ```
> 
> this command gives me some incorrect solution for error it shows,
> 
> ```rust
> error: you seem to be trying to use `&Box<T>`. Consider using just `&T`
>    --> src/internal_events/http_client.rs:100:16
>     |
> 100 |     pub error: &'a Box<dyn std::error::Error + Send + Sync>,
>     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&'a (dyn std::error::Error + Send + Sync)`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrowed_box
> ```
> ...


Did this happen when you tried to use `vector_lib::Error`? Note that type is a typedef for `Box<dyn std::error::Error + Send + Sync + 'static>`, it is already in a `Box`. So you can do the following:

```
#[derive(Debug)]
pub struct AuthExtensionError<'a> {
    pub error: &'a vector_lib::Error,
}
```

Let me know if that doesn't work, I can checkout the branch and try locally. 

> Another problem I faced is test for mtls - I wanted to make a hyper server expecting client cert but there was no crate in vector and I didn't want to pull for dependencies only for this one test.

Interested to hear more about the problem you are facing. Generally, if it's just a dev-dependency and it is a well known/used crate, it's fine to introduce it.

KowalczykBartek commented:
```
Let me know if that doesn't work, I can checkout the branch and try locally.
```
so, I can let you now it works (with your proposed change) - honestly no idea why but it compile and also ```check rust --clippy``` also passes.
Let me then construct a test based on hyper-tls, I will push it later

KowalczykBartek commented:
@pront I pushed a test for mtls scenario - it looks terrible honestly but I don't think I can do it better :) 
Question here is, where to move this https://github.com/vectordotdev/vector/pull/21583/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R189 , right now I placed it randomly to make my code running ( I don't get this cargo.toml structure yet) .

pront commented:
> @pront I pushed a test for mtls scenario - it looks terrible honestly but I don't think I can do it better :) Question here is, where to move this https://github.com/vectordotdev/vector/pull/21583/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R189 , right now I placed it randomly to make my code running ( I don't get this cargo.toml structure yet) .

Thank you! I will take a look today.

pront commented:
You can use to update licenses:

```
 dd-rust-license-tool write && git commit -am "dd-rust-license-tool write"
```

pront commented on src/http.rs:
```diff
@@ -556,10 +995,19 @@ where
 
 #[cfg(test)]
 mod tests {
```
Nit: Let's extract these into their own `tests.rs` because this file is getting too large.

KowalczykBartek replied:
what to do with method like 
```rust
fn default_request_headers<B>(request: &mut Request<B>, user_agent: &HeaderValue) {
    if !request.headers().contains_key("User-Agent") {
        request
            .headers_mut()
            .insert("User-Agent", user_agent.clone());
    }

    if !request.headers().contains_key("Accept-Encoding") {
        // hardcoding until we support compressed responses:
        // https://github.com/vectordotdev/vector/issues/5440
        request
            .headers_mut()
            .insert("Accept-Encoding", HeaderValue::from_static("identity"));
    }
}
```
make it public or copy ?

pront replied:
We could make them public. But since this PR is getting too big, feel free to ignore this comment. Resolving this thread.

pront commented on src/http.rs:
```diff
@@ -811,4 +1259,483 @@ mod tests {
+        // Each value (index) in vec, means invocation of get_now_fn by OAuth2Extension, so
+        // first call returns Duration::from_secs(11), second, Duration::from_secs(12) and so on,
+        // because of that we have full controll over time here.
+        let mocked_seconds_since_epoch = [11, 12, 20, 21, 22, 23];
+        let counter = Arc::new(Mutex::new(0));
+        let get_now_fn = move || {
+            let counter = Arc::clone(&counter);
+            let mut counter = counter.lock().unwrap();
+            let i = *counter;
+            *counter += 1;
+            Duration::from_secs(mocked_seconds_since_epoch[i])
+        };
```
I think we can use tokio test utils instead of manual implementing this: 
* https://docs.rs/tokio/latest/tokio/time/fn.advance.html 
* `#[tokio::test(start_paused = true)]`
* When you do the above, the test implementation will use a fake clock but the actual code will use the real system clock.

KowalczykBartek replied:
but how to do this when single call 
```
let second_acquisition = extension.get_token().await.unwrap();
```
is expected to have different results of same get_now() call, so I need to have n, n+1, in Java world I would mock it like this
```
when(time.getNow()).
   thenReturn(N)
   thenReturn(N+1)
   thenReturn(N+2)
   thenReturn(N+3)
```
here, I can only do advance, that's why I extracted this to function (which I can mock in tests)

pront replied:
I see. Can you do something like this?

```
#[tokio::test(start_paused = true)]
async fn test_token_acquisition() {
    let result = extension.get_token().await.unwrap();
    time::advance(Duration::from_secs(1)).await;
    // fake time advanced by 1 second

    let result = extension.get_token().await.unwrap();
    time::advance(Duration::from_secs(1)).await;
    // fake time advanced by 1 second
    
    // repeat... 
}
```

KowalczykBartek replied:
Let me explain a little bit clearer, under single call to method we want to test, so
```rust
extension.get_token().await.unwrap();
```
we have two calls to method we are mocking
```rust
SystemTime::now()
```
this is used, in two places (under the hood of this single get_token() call), first
```rust
if self.get_time_now().as_millis() < token.expires_after_ms {
```
and then
```rust
let token_will_expire_after_ms =
            OAuth2Extension::calculate_valid_until(now, self.grace_period, &token);
```

pront replied:
I see, thank you for the explanation.  Just to be clear I am not against what you have here now. But I would love to refactor this so it's a bit cleaner.I think the problem is that the current implementation of `get_token` does too many things. 

What if we split it into three functions:
```
fn is_token_expired(&self, token: &Token) -> bool { /* ... */ }
fn get_cached_token() -> Option<ExpirableToken> { /* ... */ }
fn async fn request_token(
        &self,
    ) -> Result<ExpirableToken, Box<dyn std::error::Error + Send + Sync>> { /* ... */ }
```

It will make the tests a bit more verbose but the actual code only uses it once. So it won't be too verbose.

KowalczykBartek replied:
```rust
    async fn get_token(&self) -> Result<String, vector_lib::Error> {
        if let Some(token) = self.acquire_token_from_cache() {
            return Ok(token.access_token);
        }

        //no valid token in cache (or no token at all)
        let new_token = self.request_token().await?;
        let token_to_return = new_token.access_token.clone();
        self.save_into_cache(new_token);

        Ok(token_to_return)
    }
```
isn't it split already ?

KowalczykBartek replied:
method responsible for request itself is ... let's say big
```rust
    async fn request_token(
        &self,
    ) -> Result<ExpirableToken, Box<dyn std::error::Error + Send + Sync>> {
        let mut request_body =
            format!("grant_type=client_credentials&client_id={}", self.client_id);

        // in case of oauth2 with mTLS (https://datatracker.ietf.org/doc/html/rfc8705) we only pass client_id,
        // so secret can be considered as optional.
        if let Some(client_secret) = &self.client_secret {
            let secret_param = format!("&client_secret={}", client_secret.inner());
            request_body.push_str(&secret_param);
        }

        let builder = Request::post(self.token_endpoint.clone());
        let builder = builder.header("Content-Type", "application/x-www-form-urlencoded");
        let request = builder
            .body(Body::from(request_body))
            .expect("Error creating request");

        let before = std::time::Instant::now();
        let response_result = self.client.request(request).await;
        let roundtrip = before.elapsed();

        let response = response_result.inspect_err(|error| {
            emit!(http_client::GotHttpWarning { error, roundtrip });
        })?;

        emit!(http_client::GotHttpResponse {
            response: &response,
            roundtrip
        });

        if !response.status().is_success() {
            let body_bytes = hyper::body::aggregate(response).await?;
            let body_str = std::str::from_utf8(body_bytes.chunk())?.to_string();
            return Err(Box::new(AcquireTokenError { message: body_str }));
        }

        let body = hyper::body::aggregate(response).await?;
        let token: Token = serde_json::from_reader(body.reader())?;

        let now = self.get_time_now();
        let token_will_expire_after_ms =
            OAuth2Extension::calculate_valid_until(now, self.grace_period, &token);

        Ok(ExpirableToken {
            access_token: token.access_token,
            expires_after_ms: token_will_expire_after_ms,
        })
    }
```
but comparable to 

```rust
    pub fn send(
        &self,
        mut request: Request<B>,
    ) -> BoxFuture<'static, Result<http::Response<Body>, HttpError>> {
        let span = tracing::info_span!("http");
        let _enter = span.enter();

        default_request_headers(&mut request, &self.user_agent);
        self.maybe_add_proxy_headers(&mut request);

        let client = self.client.clone();
        let auth_extension = self.auth_extension.clone();

        let fut = async move {
            if let Some(auth_extension) = auth_extension {
                let auth_span = tracing::info_span!("auth_extension");
                auth_extension
                    .modify_request(&mut request)
                    .instrument(auth_span.clone().or_current())
                    .await
                    .inspect_err(|error| {
                        // Emit the error into the internal events system.
                        emit!(http_client::AuthExtensionError { error });
                    })
                    .context(AuthenticationExtensionSnafu)?;
            }

            emit!(http_client::AboutToSendHttpRequest { request: &request });
            let response: client::ResponseFuture = client.request(request);

            // Capture the time right before we issue the request.
            // Request doesn't start the processing until we start polling it.
            let before = std::time::Instant::now();

            // Send request and wait for the result.
            let response_result = response.await;

            // Compute the roundtrip time it took to send the request and get
            // the response or error.
            let roundtrip = before.elapsed();

            // Handle the errors and extract the response.
            let response = response_result
                .inspect_err(|error| {
                    // Emit the error into the internal events system.
                    emit!(http_client::GotHttpWarning { error, roundtrip });
                })
                .context(CallRequestSnafu)?;

            // Emit the response into the internal events system.
            emit!(http_client::GotHttpResponse {
                response: &response,
                roundtrip
            });
            Ok(response)
        }
        .instrument(span.clone().or_current());

        Box::pin(fut)
    }
```

pront replied:
> isn't it split already ?

No, what I meant was to change the following:
```
let token = self.get_token().await?;
```
to:
```
let token = match self.get_cached_token() {
    Some(cached_token) if !self.is_token_expired(&cached_token) => {
        cached_token // Return the valid cached token
    }
    _ => {
        // No valid token found, refresh
        self.refresh().await?
    }
};
```

However I understand this is getting too complex and the code looks correct to me. 

1. You can do the above refactoring
2. Delete the tests that depends on fake time  and also delete `get_time_now_fn` from `OAuth2Extension`

KowalczykBartek replied:
honestly I cannot understand benefits of this move, 
```rust
let token = match self.get_cached_token() {
    Some(cached_token) if !self.is_token_expired(&cached_token) => {
        cached_token // Return the valid cached token
    }
    _ => {
        // No valid token found, refresh
        self.refresh().await?
    }
};
```
this is the same amount of logic, moved one step lower in the invocation tree.
```
Delete the tests that depends on fake time and also delete get_time_now_fn from OAuth2Extension
```
I can leave this test, there will be just no part of the test for token refreshing mechanism (so, the same scope as with mtls test)

pront replied:
> there will be just no part of the test for token refreshing mechanism (so, the same scope as with mtls test - because get_time_now_fn will be removed), is that ok ?

Yes that's absolutely fine 👍

KowalczykBartek replied:
fixed.

pront reviewed (approved):
Left some ideas for improvement but overall this looks good

KowalczykBartek commented:
@pront can you please suggest where should i move those dependencies https://github.com/vectordotdev/vector/pull/21583/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R189 ?

pront commented:
> @pront can you please suggest where should i move those dependencies https://github.com/vectordotdev/vector/pull/21583/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R189 ?

These are test only deps right? If yes, they should go under `[dev-dependencies]`.

KowalczykBartek commented on src/http.rs:
```diff
@@ -72,31 +85,297 @@ impl HttpError {
+#[derive(Debug, Deserialize)]
+struct Token {
+    access_token: String,
+    expires_in: u32,
```
fixed

pront commented on src/http.rs:
```diff
@@ -72,31 +85,298 @@ impl HttpError {
+#[derive(Debug, Deserialize)]
+struct Token {
+    access_token: String,
+    #[serde(rename = "expires_in")] 
```
Ah sorry, was this a reserved name for an interface?

KowalczykBartek replied:
this is the response from oauth2 server, so I need to remap it

pront replied:
Ah I see, what you did makes sense 👍

KowalczykBartek replied:
so, can I rename it back to expires_in ?

KowalczykBartek replied:
I pushed revert to expires_in

KowalczykBartek replied:
I read comment once again - I guess you wanted to put a comment about a time unit behind this property ? if yes, one is already on place 
```rust
const fn calculate_valid_until(now: Duration, grace_period: u32, token: &Token) -> u128 {
    // 'expires_in' means, in seconds, for how long it will be valid, lets say 5min,
    // to not cause some random 4xx, because token expired in the meantime, we will make some
    // room for token refreshing, this room is a grace_period.
    let (mut grace_period_seconds, overflow) = token.expires_in.overflowing_sub(grace_period);
```
I also added explicit comment here
```rust
#[derive(Debug, Deserialize)]
struct Token {
    access_token: String,
    // This property, according to RFC, is expected to be in seconds.
    expires_in: u32,
}
```

pront replied:
This is good  now 👍

pront commented:
Thank you again @KowalczykBartek for this great PR. You went above and beyond 👍 I enqueued it for merging.

KowalczykBartek commented:
Thank you @pront for support and patience :)

KowalczykBartek commented:
> One more thing here, I noticed this failure: https://github.com/vectordotdev/vector/actions/runs/11728482979/job/32672282844?pr=21583.
> 
> This might be helpful:
> 
> ```
> make generate-component-docs
> # review any changes
> # commit if they look good
> make check-component-docs
> git push
> ```

hmm, generate-component-docs fail with 
```
vector/scripts/generate-component-docs.rb:1753:in `<main>': undefined method `filter_map' for #<Hash:0x00000001380c15a0> (NoMethodError)
```

jszwedko commented:
> hmm, generate-component-docs fail with
> 
> ```
> vector/scripts/generate-component-docs.rb:1753:in `<main>': undefined method `filter_map' for #<Hash:0x00000001380c15a0> (NoMethodError)
> ```

I think that means you are using Ruby 2, that script requires Ruby 3 (that should really be clearer 😓 ).

KowalczykBartek commented:
you are right, I have ruby 2, let me try to fix this :D

KowalczykBartek commented:
I was able to run this command, 
```
# review any changes
# commit if they look good
```
but this this is tricky, I am not sure what this http.cue is :) It looks, good, but is it good, not sure :)

jszwedko commented on .github/workflows/semantic.yml:
```diff
@@ -29,6 +29,7 @@ jobs:
             revert
 
           scopes: |
+            http
```
```suggestion
```

This change seems to affect the `http` and `axiom` sinks? Could we use those pre-existing scopes instead?

KowalczykBartek replied:
I don't know :) I did this because it was one of the review comments https://github.com/vectordotdev/vector/pull/21583#issuecomment-2450699285 :)

pront replied:
I thought it is worth introducing a scope since this affects a lower level module used by multiple components. But I don't really have a strong opinion here.

jszwedko replied:
Aha gotcha, sorry for the run-around 😅 Maybe `http provider` would be more consistent? We have other `x providers` already as scopes to use for lower level changes that affect multiple components that share a "provider".

KowalczykBartek replied:
I removed this http component from list as suggested

KowalczykBartek commented:
Windows has some differences across error messages, so, I just remove error msg expectation for connection refused and internal server errors, its too much library and platform dependent

singhbaljit commented:
Chiming in a bit late... this is highly appreciated! I'll certainly test this feature and report any issues I find. Thank you very much!

jszwedko commented:
> > The next Vector release, probably v0.43 will include this enhancement. Current version is https://vector.dev/releases/0.42.0/.
> 
> @pront when is v0.43 expected to be released?

Dec 2nd. You can see this via the [milestone](https://github.com/vectordotdev/vector/milestone/135). Also calendar.vector.dev which is a public calendar with the releases on it (without version numbers).

pront commented:
Hi all, unfortunately I found a UX issue with this PR which should be fixed before including in a Vector release. Apologies, I missed it when reviewing because I was focusing on other technical details. 

Note how we already provide: 
1. https://vector.dev/docs/reference/configuration/sinks/http/#auth
2. https://vector.dev/docs/reference/configuration/sinks/http/#tls
3. In code: https://github.com/vectordotdev/vector/blob/master/src/http.rs#L677-L710 (note how it is tagged with `strategy`)


I created the website locally: 
<img width="307" alt="image" src="https://github.com/user-attachments/assets/0cb44668-41b1-47a1-a981-1f48d3acfd59">

We would have seen this if the PR branch contained `website` because the CI would generate a preview.

There is a lot of redundancy in the newly added `authorization_config`. The newly added `oath_2` could have simply been a new variant under `Auth`. This is backwards compatible as well since we are just adding a new strategy. 

So we have two options here:
1. follow-up and fix the UX before the next release
2. revert the PR

Is option 1 something that you are willing to contribute?

KowalczykBartek commented:
> I missed it when reviewing because I was focusing on other technical details.

Forrest Gump used to say: 'Shit happens... sometimes' :)

> Is option 1 something that you are willing to contribute?

maybe not willing, but reverting it back would be sad for me :) 

I opened new PR, I cannot push here,  https://github.com/vectordotdev/vector/pull/21857 
as i understand, you opt for this 
```rust

/// Configuration of the authentication strategy for HTTP requests.
///
/// HTTP authentication should be used with HTTPS only, as the authentication credentials are passed as an
/// HTTP header without any additional encryption beyond what is provided by the transport itself.
#[configurable_component]
#[derive(Clone, Debug, Eq, PartialEq)]
#[serde(deny_unknown_fields, rename_all = "snake_case", tag = "strategy")]
#[configurable(metadata(docs::enum_tag_description = "The authentication strategy to use."))]
pub enum Auth {
    /// Basic authentication.
    ///
    /// The username and password are concatenated and encoded via [base64][base64].
    ///
    /// [base64]: https://en.wikipedia.org/wiki/Base64
    Basic {
        /// The basic authentication username.
        #[configurable(metadata(docs::examples = "${USERNAME}"))]
        #[configurable(metadata(docs::examples = "username"))]
        user: String,

        /// The basic authentication password.
        #[configurable(metadata(docs::examples = "${PASSWORD}"))]
        #[configurable(metadata(docs::examples = "password"))]
        password: SensitiveString,
    },

    /// Bearer authentication.
    ///
    /// The bearer token value (OAuth2, JWT, etc.) is passed as-is.
    Bearer {
        /// The bearer authentication token.
        token: SensitiveString,
    },

    /// Authentication based on OAuth 2.0 protocol.
    ///
    /// This strategy allows to dynamically acquire and use token based on provided parameters.
    /// Both standard client_credentials and mTLS extension is supported, for standard client_credentials just provide both
    /// client_id and client_secret parameters:
    ///
    /// # Example
    ///
    /// ```yaml
    /// strategy:
    ///  strategy: "o_auth2"
    ///  client_id: "client.id"
    ///  client_secret: "secret-value"
    ///  token_endpoint: "https://yourendpoint.com/oauth/token"
    /// ```
    /// In case you want to use mTLS extension [rfc8705](https://datatracker.ietf.org/doc/html/rfc8705), provide desired key and certificate,
    /// together with client_id (with no client_secret parameter).
    ///
    /// # Example
    ///
    /// ```yaml
    /// strategy:
    ///  strategy: "o_auth2"
    ///  client_id: "client.id"
    ///  token_endpoint: "https://yourendpoint.com/oauth/token"
    /// tls:
    ///  crt_path: cert.pem
    ///  key_file: key.pem
    /// ```
    OAuth2 {
        /// Token endpoint location, required for token acquisition.
        #[configurable(metadata(docs::examples = "https://auth.provider/oauth/token"))]
        token_endpoint: String,

        /// The client id.
        #[configurable(metadata(docs::examples = "client_id"))]
        client_id: String,

        /// The sensitive client secret.
        #[configurable(metadata(docs::examples = "client_secret"))]
        client_secret: Option<SensitiveString>,

        /// The grace period configuration for a bearer token.
        /// To avoid random authorization failures caused by expired token exception,
        /// we will acquire new token, some time (grace period) before current token will be expired,
        /// because of that, we will always execute request with fresh enough token.
        #[serde(default = "default_oauth2_token_grace_period")]
        #[configurable(metadata(docs::examples = 300))]
        #[configurable(metadata(docs::type_unit = "seconds"))]
        #[configurable(metadata(docs::human_name = "Grace period for bearer token."))]
        grace_period: u32,
    },
}
```
